## Introduction to Python for Machine Learning
#### Hamoye #1

#### Import libraries & load the data




In [ ]:
import numpy as np
import pandas as pd


In [ ]:
data = pd.read_csv('https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv')
data.head()

record_id  ...  fuel_cost_per_mmbtu
0   f1_fuel_1994_12_1_0_7  ...                1.121
1  f1_fuel_1994_12_1_0_10  ...                1.120
2   f1_fuel_1994_12_2_0_1  ...                1.650
3   f1_fuel_1994_12_2_0_7  ...                1.970
4  f1_fuel_1994_12_2_0_10  ...                2.570

[5 rows x 11 columns]

#### Fuel type with the lowest average of fuel cost per unit burned

In [ ]:
fuel = data[['fuel_type_code_pudl', 'fuel_cost_per_unit_burned']]
fuel.set_index('fuel_type_code_pudl')
fuel = fuel.groupby(["fuel_type_code_pudl"]).mean()
fuel = fuel.sort_values(by=['fuel_cost_per_unit_burned'], ascending=True)
fuel.index[0]

'gas'

###### **Gas** is the ype of fuel with the lowest average of fuel cost per unit burned.

#### Standard Deviation and 75th percentile of energy per unit


In [ ]:
data.describe()

utility_id_ferc1  ...  fuel_cost_per_mmbtu
count      29523.000000  ...         29523.000000
mean         118.601836  ...            19.304354
std           74.178353  ...          2091.540939
min            1.000000  ...           -41.501000
25%           55.000000  ...             1.940000
50%          122.000000  ...             4.127000
75%          176.000000  ...             7.745000
max          514.000000  ...        359278.000000

[8 rows x 7 columns]

##### The standard deviation is **10.60** and the 75th percentile is **17.00**

#### The skewness and kurtosis for the fuel quantity burned 

In [ ]:
data.kurtosis()

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64

In [ ]:
data.skew()

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64

##### The kurtosis is **651.37** and the skewness is **15.85**

#### The feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows

In [ ]:
data.isnull().sum()

record_id                       0.000000
utility_id_ferc1                0.000000
report_year                     0.000000
plant_name_ferc1                0.000000
fuel_type_code_pudl             0.000000
fuel_unit                       0.609694
fuel_qty_burned                 0.000000
fuel_mmbtu_per_unit             0.000000
fuel_cost_per_unit_burned       0.000000
fuel_cost_per_unit_delivered    0.000000
fuel_cost_per_mmbtu             0.000000
dtype: float64

In [ ]:
data.shape

(29523, 11)

In [ ]:
percentage = (180*100)/29523
percentage

0.6096941367747181

In [ ]:
data.isnull().sum()* 100 / len(data)

record_id                       0.000000
utility_id_ferc1                0.000000
report_year                     0.000000
plant_name_ferc1                0.000000
fuel_type_code_pudl             0.000000
fuel_unit                       0.609694
fuel_qty_burned                 0.000000
fuel_mmbtu_per_unit             0.000000
fuel_cost_per_unit_burned       0.000000
fuel_cost_per_unit_delivered    0.000000
fuel_cost_per_mmbtu             0.000000
dtype: float64

##### The feature is: **fuel_unit**, the missing values: **180** and the percentage: **0.609** or **0.610**

#### The features has the second and third lowest correlation with the Fuel Cost Per Unit Burned

In [ ]:
 corr = data[['fuel_qty_burned','fuel_mmbtu_per_unit','fuel_cost_per_unit_burned','fuel_cost_per_unit_delivered','fuel_cost_per_mmbtu']]
 corr.head()

fuel_qty_burned  ...  fuel_cost_per_mmbtu
0        5377489.0  ...                1.121
1       10486945.0  ...                1.120
2        2978683.0  ...                1.650
3        3739484.0  ...                1.970
4          40533.0  ...                2.570

[5 rows x 5 columns]

In [ ]:
corr_one = data['fuel_cost_per_unit_burned'].corr(data['fuel_qty_burned'])
corr_one

-0.018535416794825225

In [ ]:
corr_two = data['fuel_cost_per_unit_burned'].corr(data['fuel_mmbtu_per_unit'])
corr_two

-0.010033808428853071

In [ ]:
corr_three = (data['fuel_cost_per_unit_delivered']).corr(data['fuel_cost_per_unit_burned'])
corr_three

0.01100740329916511

In [ ]:
corr_four = data['fuel_cost_per_unit_burned'].corr(data['fuel_cost_per_mmbtu'])
corr_four

-0.0004372104439704003

In [ ]:
corr_five = data['fuel_cost_per_unit_burned'].corr(data['report_year'])
corr_five

0.013599134438134116

In [ ]:
corr_six = data['fuel_cost_per_unit_burned'].corr(data['utility_id_ferc1'])
corr_six

-0.0378627075999726

In [ ]:
corr = pd.Series(data=[corr_one, corr_two, corr_three, corr_four, corr_five, corr_six])
corr = corr.sort_values(ascending=True)
corr

5   -0.037863
0   -0.018535
1   -0.010034
3   -0.000437
2    0.011007
4    0.013599
dtype: float64

##### The second lowest feature correlation is **fuel_qty_burned** and the third lowest feature is **fuel_mmbtu_per_unit** 

#### For the fuel type coal, the percentage change in the fuel cost per unit burned in 1998 compared to 1994

In [ ]:
change = data[['report_year', 'fuel_type_code_pudl', 'fuel_cost_per_unit_burned']]
change = change.set_index(['report_year', 'fuel_type_code_pudl'])
change = change.groupby(by=['report_year', 'fuel_type_code_pudl']).sum()
change = change.reset_index()

fuel = change[change['fuel_type_code_pudl'] == 'coal']
change = 11902.597 - 14984.572
(change * 100) / 14984.572

-20.567654518260518

##### The change is **20.56%** on 1998 compared to 1994

#### Year with the Highest average of fuel cost per unit delivered

In [128]:
avg = data[['report_year', 'fuel_cost_per_unit_delivered']]
avg = avg.groupby(["report_year"]).mean()
avg = avg.sort_values(by='fuel_cost_per_unit_delivered', ascending=True)
avg

fuel_cost_per_unit_delivered
report_year                              
1999                            25.551627
1995                            32.735269
2006                            38.657484
2005                            41.438184
2007                            43.325023
2017                            46.196861
2002                            47.594361
2003                            55.663493
2008                            58.588197
2011                            59.774667
2001                            60.050396
2012                            60.994502
1994                            63.636060
2010                            91.862105
2016                           103.901761
2004                           139.524275
2013                           172.307591
2014                           192.737183
1998                           287.154420
2015                           326.535511
2018                           499.269966
2009                           652.694163
2000                           985.362877
1996                          9196.705948
1997                         11140.197239

##### The year is **1997** with the highest average